### PR TITLE
server : Add option to return token pieces in /tokenize endpoint

### PR DESCRIPTION
**Pull request by mathijshenquet**
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

## Description

This PR enhances the `/tokenize` endpoint by adding an option to return token pieces along with their IDs. This feature allows users to easily understand what each token represents without needing to make additional API calls or perform client-side lookups.

## Motivation

Often, users want to know not just the token IDs but also what these tokens represent in the original text. Previously, this required either inefficient additional API calls or complex client-side logic. This change makes it much easier and more efficient to get this information in a single request.

## Changes

1. Added a new `with_pieces` boolean parameter to the `/tokenize` endpoint.
2. When `with_pieces` is true, the response includes both token IDs and their corresponding pieces.
3. Updated the endpoint documentation to reflect the new parameter and response format.
4. The change is backward compatible: if `with_pieces` is not specified or is false, the endpoint behaves as before.

## Testing

- Added unit tests to cover the new functionality.
- Manually tested the endpoint with various inputs and parameter combinations.

## Documentation

Updated the API documentation to include the new `with_pieces` parameter and to show example responses for both cases (with and without pieces).

NB: I wasn't able to run the CI pipeline locally as I'm currently on Windows.

### CODE REVIEW

**Review by ggerganov — APPROVED** [review]

File: examples/server/tests/features/steps/steps.py [hunk 1]
```diff
@@ -685,6 +685,32 @@ def step_tokenize_set_add_special(context):
     context.tokenize_add_special = True
 
 
+@step("tokenizing with pieces")
+@async_run_until_complete
```
Comment (ngxson) [inline examples/server/tests/features/steps/steps.py:689]: You must also add these step into .feature file. Otherwise, they will not be called

**Review by ngxson — CHANGES_REQUESTED** [review]

**Comment by ngxson** [comment]
Also, this will fail if one of the pieces has incomplete unicode bytes. For example:

```
{"content": "💁", "with_pieces": true}
```

Response:

```
{
    "error": {
        "code": 500,
        "message": "[json.exception.type_error.316] incomplete UTF-8 string; last byte: 0x92",
        "type": "server_error"
    }
}
```

File: examples/server/tests/features/steps/steps.py (hunk 1 above)
Comment (mathijshenquet) [reply examples/server/tests/features/steps/steps.py:689]: Thanks, I have added them

**Comment by mathijshenquet** [comment]
> Also, this will fail if one of the pieces has incomplete unicode bytes. For example:

For the case that a piece is invalid utf8 I have added a fallback where a list of bytes will be sent instead.

**Comment by mathijshenquet** [comment]
ping @ngxson

**Review by ngxson — APPROVED** [review]
LGTM. Thanks. Let's merge after the CI passes

Small suggestion: we actually already had a version of `is_valid_utf8` in `process_token`, would be nice if we can deduplicate the code by reusing that. It can be done in a follow-up PR though.

**Comment by ggerganov** [comment]
@mathijshenquet Sorry for the delay. Let's resolve the conflict and merge.

**Comment by ngxson** [comment]
I merged with upstream master. That should fix the CI problem. (Let's merge once CI passes)

**Pull request merged**